File: Lithuaningo.Frontend/src/services/data/wordService.ts
// services/firebase/wordService.ts
import { COLLECTIONS } from "@config/constants";
import firestore from "@react-native-firebase/firestore";

export interface WordForm {
  lithuanian: string;
  english: string;
}

export interface Word {
  id: string; // Document ID, which is the base form of the word
  wordForms: WordForm[]; // All known forms of the word along with translations
  englishTranslation: string; // Translation of the base form in English
  imageUrl: string; // URL of the image related to the word
  additionalInfo?: string; // Optional additional information
}

const fetchWords = async (): Promise<Word[]> => {
  const snapshot = await firestore().collection(COLLECTIONS.WORDS).get();
  const wordsData = snapshot.docs.map((doc) => ({
    id: doc.id,
    ...doc.data(),
  })) as Word[];
  return wordsData;
};

const addWordForReview = async (wordData: Word): Promise<void> => {
  try {
    await firestore().collection(COLLECTIONS.PENDING_WORDS).add(wordData);
  } catch (error) {
    console.error("Error adding word for review:", error);
    throw error;
  }
};

const addMissingWord = async (word: string): Promise<void> => {
  try {
    const docRef = firestore().collection(COLLECTIONS.MISSING_WORDS).doc(word);
    const doc = await docRef.get();

    if (!doc.exists) {
      await docRef.set({ id: word });
    }
  } catch (error) {
    console.error("Error adding missing word:", error);
    throw error;
  }
};

export default {
  fetchWords,
  addWordForReview,
  addMissingWord,
};

File: Lithuaningo.API/Services/WordService.cs
using Google.Cloud.Firestore;

public class WordService
{
    private readonly FirestoreDb _db;

    public WordService(FirestoreDb db)
    {
        _db = db;
    }

    public async Task<List<Word>> FetchWordsAsync()
    {
        var snapshot = await _db.Collection("words").GetSnapshotAsync();
        return snapshot.Documents.Select(doc => doc.ConvertTo<Word>()).ToList();
    }

    public async Task AddWordForReviewAsync(Word word)
    {
        await _db.Collection("pending_words").AddAsync(word);
    }

    public async Task AddMissingWordAsync(string word)
    {
        var docRef = _db.Collection("missing_words").Document(word);
        var doc = await docRef.GetSnapshotAsync();
        if (!doc.Exists)
        {
            await docRef.SetAsync(new { Id = word });
        }
    }
}